lib/Base/Organization/Collection.vala

```diff
@@ -16,192 +16,123 @@
    */
-  public abstract async void pull_posts () throws Error;
+  public Type get_item_type () {
+    return typeof (T);
+  }
 
   /**
-   * Checks if a post in the list replies to the post previous to it.
+   * Get the number of items in this list.
    *
-   * @param post The post to check for.
-   * @param parent If true, a reference to the parent post is set.
-   *
-   * @return If the post replies to the post previous to it.
+   * @return The number of items in this list.
    */
-  public bool connected_to_previous (Post post, out Post? parent = null) {
-    // Retrieve the post list as ListStore
-    var store = post_list as ListStore;
-    if (store == null) {
-      return false;
-    }
-
-    // Check if the post exists in the list
-    uint index;
-    if (! store.find (post, out index)) {
-      return false;
-    }
-
-    // Find the post before the parameter
-    var prev_post = store.get_item (index - 1) as Post;
-    if (prev_post != null) {
-      // Check if prev_post is parent
-      bool has_parent = post.replied_to_id == prev_post.id;
-      // Return the result
-      parent = has_parent ? prev_post : null;
-      return has_parent;
-    }
-    return false;
+  public uint get_n_items () {
+    return (uint) items.get_length;
   }
 
   /**
-   * Checks if a post in the list replies to the post next to it.
+   * Returns the nth item in this list.
    *
-   * @param post The post to check for.
-   * @param child If true, a reference to the child post is set.
+   * @param position The position to look for.
    *
-   * @return If the post replies to the post next to it.
+   * @return The item at the position, or null if position is invalid.
    */
-  public bool connected_to_next (Post post, out Post? child = null) {
-    // Retrieve the post list as ListStore
-    var store = post_list as ListStore;
-    if (store == null) {
-      return false;
+  public Object? get_item (uint position) {
+    SequenceIter<T>? iter = null;
+
+    // Check if we can access the item quickly from the cache
+    if (last_position_valid) {
+      if (last_position == position) {
+        iter = last_iterator;
+      } else if (last_position == position - 1) {
+        iter = last_iterator.next ();
+      } else if (last_position == position + 1) {
+        iter = last_iterator.prev ();
+      }
     }
 
-    // Check if the post exists in the list
-    uint index;
-    if (! store.find (post, out index)) {
-      return false;
+    if (iter == null) {
+      iter = items.get_iter_at_pos ((int) position);
     }
 
-    // Find the post before the parameter
-    var next_post = store.get_item (index + 1) as Post;
-    if (next_post != null) {
-      // Check if next_post is child
-      bool has_child = next_post.replied_to_id == post.id;
-      // Return the result
-      child = has_child ? next_post : null;
-      return has_child;
-    }
-    return false;
+    // Update cache position
+    last_position_valid = true;
```

Comment:
What about if the position is beyond the end of the collection? Is it still valid then? 🙂

Because I _think_ this would result in a following call giving unspecified behaviour, because we don't do the `iter.is_end()` check in the "check the cache" block.

Reply:
That was pretty much taken from the code from `Gio.ListStore`. So, unless I've made an error while adapting this, I guess it should work?

Reply:
No, you're right. The `if (last_position_valid){…}` block doesn't return early, it just gets the iterator, so it would be fine. The `iter.is_end()` check _would_ run on the next go. I wasn't following the code closely enough and thought that we'd end up with a "bad" iterator that we'd then call without a guard on the next call.

And yeah, it [looks like it matches](https://gitlab.gnome.org/GNOME/glib/-/blob/main/gio/gliststore.c#L201-229).